Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+            int batchSize;  // Batches of cache entries to be deleted.
+            var now = DateTime.UtcNow;
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderByDescending("AbsoluteExpiration")
+                    .StartAfter(now)
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    if (docSnapshot.ConvertTo<CacheDoc>()
+                        .AbsoluteExpiration.HasValue)
+                    {
+                        writeBatch.Delete(docSnapshot.Reference,
+                            Precondition.LastUpdated(
+                                docSnapshot.UpdateTime.GetValueOrDefault()));
+                        batchSize += 1;
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize == pageSize);
+
+            // Purge entries whose SlidingExpiration has passed.
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderBy("LastRefresh")
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    CacheDoc doc = docSnapshot.ConvertTo<CacheDoc>();
+                    if (doc.SlidingExpirationSeconds.HasValue)
+                    {
+                        var slidingExpiration =
+                            doc.LastRefresh.GetValueOrDefault()
+                            + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.Value);
+                        if (slidingExpiration < now)
+                        {
+                            writeBatch.Delete(docSnapshot.Reference,
+                                Precondition.LastUpdated(
+                                    docSnapshot.UpdateTime.GetValueOrDefault()));
+                            batchSize += 1;
+                        }
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;

```

Comment:
Just return rather than throwing an exception? Note that `CommitAsync` *will* throw an exception if cancellation is requested. It seems odd to do different things in those cases.